# Actividad - Proyecto práctico: DQN con SpaceInvaders

**Máster en Inteligencia Artificial - Aprendizaje por Refuerzo**

> La actividad se desarrollará en grupos pre-definidos de 2-3 alumnos. Se debe indicar los nombres en orden alfabético (de apellidos). Recordad que esta actividad se corresponde con un 30% de la nota final de la asignatura. Se debe entregar entregar el trabajo en la presente notebook.
*   Alumno 1: Manjarrez Jinez Paúl Alexander
*   Alumno 2: Bravo Esquivias Gustavo Roger
*   Alumno 3: Pardo Hortua Eduar Uriel
*   Alumno 4: Basurto Cruz Edgar Daniel

Link de GitHub: https://github.com/edgarbasurto/08MIAR_10_B_2024-25_Aprendizaje_por_refuerzo

---
## **PARTE 1** - Instalación y requisitos previos

> Las prácticas han sido preparadas para poder realizarse en el entorno de trabajo de Google Colab. Sin embargo, esta plataforma presenta ciertas incompatibilidades a la hora de visualizar la renderización en gym. Por ello, para obtener estas visualizaciones, se deberá trasladar el entorno de trabajo a local. Por ello, el presente dosier presenta instrucciones para poder trabajar en ambos entornos. Siga los siguientes pasos para un correcto funcionamiento:
1.   **LOCAL:** Preparar el enviroment, siguiendo las intrucciones detalladas en la sección *1.1.Preparar enviroment*.
2.  **AMBOS:** Modificar las variables "mount" y "drive_mount" a la carpeta de trabajo en drive en el caso de estar en Colab, y ejecturar la celda *1.2.Localizar entorno de trabajo*.
3. **COLAB:** se deberá ejecutar las celdas correspondientes al montaje de la carpeta de trabajo en Drive. Esta corresponde a la sección *1.3.Montar carpeta de datos local*.
4.  **AMBOS:** Instalar las librerías necesarias, siguiendo la sección *1.4.Instalar librerías necesarias*.


---
### 1.1. Preparar enviroment (solo local)



> Para preparar el entorno de trabajo en local, se han seguido los siguientes pasos:
1. En Windows, puede ser necesario instalar las C++ Build Tools. Para ello, siga los siguientes pasos: https://towardsdatascience.com/how-to-install-openai-gym-in-a-windows-environment-338969e24d30.
2. Instalar Anaconda
3. Siguiendo el código que se presenta comentado en la próxima celda: Crear un enviroment, cambiar la ruta de trabajo, e instalar librerías básicas.


```
conda create --name miar_rl python=3.8
conda activate miar_rl
cd "PATH_TO_FOLDER"
conda install git
pip install jupyter
```


4. Abrir la notebook con *jupyter-notebook*.



```
jupyter-notebook
```


---
### 1.2. Localizar entorno de trabajo: Google colab o local

In [ ]:
# ATENCIÓN!! Modificar ruta relativa a la práctica si es distinta (drive_root)
mount='/content/gdrive'
drive_root = mount + "/My Drive/08_MIAR/actividades/proyecto practico"

try:
    from google.colab import drive
    IN_COLAB=True
except:
    IN_COLAB=False

---
### 1.3. Montar carpeta de datos local (solo Colab)

In [ ]:
# Switch to the directory on the Google Drive that you want to use
import os
if IN_COLAB:
  print("We're running Colab")

  if IN_COLAB:
    # Mount the Google Drive at mount
    print("Colab: mounting Google drive on ", mount)

    drive.mount(mount)

    # Create drive_root if it doesn't exist
    create_drive_root = True
    if create_drive_root:
      print("\nColab: making sure ", drive_root, " exists.")
      os.makedirs(drive_root, exist_ok=True)

    # Change to the directory
    print("\nColab: Changing directory to ", drive_root)
    %cd $drive_root
# Verify we're in the correct working directory
%pwd
print("Archivos en el directorio: ")
print(os.listdir())

Archivos en el directorio: 
['Proyecto_DQN.ipynb', '.ipynb_checkpoints']


---
### 1.4. Instalar librerías necesarias

In [ ]:
if IN_COLAB:
    %pip install gym==0.17.3
    %pip install git+https://github.com/Kojoley/atari-py.git
    %pip install keras-rl2==1.0.5
    %pip install tensorflow==2.18
else:
    %pip install torch==2.7.1
    %pip install torchvision
    %pip install torchaudio
    %pip install gymnasium[atari,accept-rom-license]==0.29
    %pip install gymnasium-atari==0.29
    %pip install ale-py==0.10
    %pip install autorom==0.6
    %pip install imageio
    %pip install opencv-python
    %pip install tqdm
    %pip install numpy

Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
zsh:1: no matches found: gymnasium[atari,accept-rom-license]==0.29
Note: you may need to restart the kernel to use updated packages.
ERROR: Could not find a version that satisfies the requirement gymnasium-atari==0.29 (from versions: none)
ERROR: No matching distribution found for gymnasium-atari==0.29
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.
Note: you may need to restart the kernel to use updated packages.


---
## **PARTE 2**. Enunciado

Consideraciones a tener en cuenta:

- El entorno sobre el que trabajaremos será _SpaceInvaders-v0_ y el algoritmo que usaremos será _DQN_.

- Para nuestro ejercicio, el requisito mínimo será alcanzado cuando el agente consiga una **media de recompensa por encima de 20 puntos en modo test**. Por ello, esta media de la recompensa se calculará a partir del código de test en la última celda del notebook.

Este proyecto práctico consta de tres partes:

1.   Implementar la red neuronal que se usará en la solución
2.   Implementar las distintas piezas de la solución DQN
3.   Justificar la respuesta en relación a los resultados obtenidos

**Rúbrica**: Se valorará la originalidad en la solución aportada, así como la capacidad de discutir los resultados de forma detallada. El requisito mínimo servirá para aprobar la actividad, bajo premisa de que la discusión del resultado sera apropiada.

IMPORTANTE:

* Si no se consigue una puntuación óptima, responder sobre la mejor puntuación obtenida.
* Para entrenamientos largos, recordad que podéis usar checkpoints de vuestros modelos para retomar los entrenamientos. En este caso, recordad cambiar los parámetros adecuadamente (sobre todo los relacionados con el proceso de exploración).
* Se deberá entregar unicamente el notebook y los pesos del mejor modelo en un fichero .zip, de forma organizada.
* Cada alumno deberá de subir la solución de forma individual.

---
## **PARTE 3**. Desarrollo y preguntas

#### Importar librerías

In [ ]:
import ale_py
import gymnasium as gym
import torch
import torch.nn as nn
import torch.nn.functional as F
import numpy as np
import random
from collections import deque


### 1. Implementación de la red neuronal

In [ ]:
# Definición de red neuronal
class DQN(nn.Module):
    def __init__(self, input_shape, num_actions):
        super(DQN, self).__init__()
        self.conv1 = nn.Conv2d(input_shape[0], 32, kernel_size=8, stride=4)
        self.conv2 = nn.Conv2d(32, 64, kernel_size=4, stride=2)
        self.conv3 = nn.Conv2d(64, 64, kernel_size=3, stride=1)

        conv_out_size = self._get_conv_out(input_shape)
        self.fc1 = nn.Linear(conv_out_size, 512)
        self.fc2 = nn.Linear(512, num_actions)

    def _get_conv_out(self, shape):
        o = torch.zeros(1, *shape)
        o = self.conv3(self.conv2(self.conv1(o)))
        return int(np.prod(o.size()))

    def forward(self, x):
        x = x / 255.0
        x = F.relu(self.conv1(x))
        x = F.relu(self.conv2(x))
        x = F.relu(self.conv3(x))
        x = torch.flatten(x, 1)
        x = F.relu(self.fc1(x))
        return self.fc2(x)

In [ ]:
# Replay buffer y frame stack
class ReplayBuffer:
    def __init__(self, capacity):
        self.buffer = deque(maxlen=capacity)

    def push(self, s, a, r, s_, d):
        self.buffer.append((s, a, r, s_, d))

    def sample(self, batch_size):
        batch = random.sample(self.buffer, batch_size)
        states, actions, rewards, next_states, dones = zip(*batch)
        return (np.stack(states), actions, rewards, np.stack(next_states), dones)

    def __len__(self):
        return len(self.buffer)

import cv2
class FrameStack:
    def __init__(self, k):
        self.k = k
        self.frames = deque([], maxlen=k)

    def reset(self, obs):
        f = self.preprocess(obs)
        for _ in range(self.k): self.frames.append(f)
        return np.stack(self.frames, axis=0)

    def step(self, obs):
        f = self.preprocess(obs)
        self.frames.append(f)
        return np.stack(self.frames, axis=0)

    def preprocess(self, obs):
        gray = cv2.cvtColor(obs, cv2.COLOR_RGB2GRAY)
        resized = cv2.resize(gray, (84, 84), interpolation=cv2.INTER_AREA)
        return resized.astype(np.uint8)

In [ ]:
# Función de acción y entrenamiento
def select_action(model, state, epsilon, num_actions, device):
    if random.random() < epsilon:
        return random.randint(0, num_actions - 1)
    state = torch.tensor(state, dtype=torch.float32).unsqueeze(0).to(device)
    with torch.no_grad():
        q = model(state)
    return q.argmax().item()

def train(model, target, buffer, opt, batch_size, gamma, device):
    if len(buffer) < batch_size: return
    s, a, r, s_, d = buffer.sample(batch_size)
    s = torch.tensor(s, dtype=torch.float32).to(device)
    a = torch.tensor(a).to(device)
    r = torch.tensor(r).to(device)
    s_ = torch.tensor(s_, dtype=torch.float32).to(device)
    d = torch.tensor(d).to(device)
    q = model(s).gather(1, a.unsqueeze(1)).squeeze()
    q_ = target(s_).max(1)[0]
    # Convert boolean done tensor to float before subtraction
    target_q = r + gamma * q_ * (1.0 - d.float())
    loss = nn.MSELoss()(q, target_q)
    opt.zero_grad()
    loss.backward()
    opt.step()

In [ ]:
# Entrenamiento del agente (resumen, se puede ajustar)
import gymnasium as gym
env = gym.make('ALE/SpaceInvaders-v5')
num_actions = env.action_space.n
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

policy = DQN((4, 84, 84), num_actions).to(device)
target = DQN((4, 84, 84), num_actions).to(device)
target.load_state_dict(policy.state_dict())
opt = torch.optim.Adam(policy.parameters(), lr=1e-4)
buffer = ReplayBuffer(100000)
stacker = FrameStack(4)

obs, _ = env.reset()
print("Shape de observación:", obs.shape)

Shape de observación: (210, 160, 3)


### 2. Implementación de la solución DQN

In [ ]:
# Entrenamiento del agente DQN
num_episodes = 500
batch_size = 32
gamma = 0.99
epsilon_start = 1.0
epsilon_final = 0.1
epsilon_decay = 100000
target_update_freq = 1000
train_freq = 4
checkpoint_path = 'modelo_final.pth'
steps_done = 0
epsilon = epsilon_start
rewards = []
from tqdm import trange

for episode in trange(num_episodes):
    obs, _ = env.reset()
    state = stacker.reset(obs)
    done = False
    total_reward = 0

    while not done:
        epsilon = epsilon_final + (epsilon_start - epsilon_final) * np.exp(-1.0 * steps_done / epsilon_decay)
        action = select_action(policy, state, epsilon, num_actions, device)
        next_obs, reward, terminated, truncated, _ = env.step(action)
        next_state = stacker.step(next_obs)
        done = terminated or truncated
        buffer.push(state, action, reward, next_state, done)
        state = next_state
        total_reward += reward
        steps_done += 1

        if steps_done % train_freq == 0:
            train(policy, target, buffer, opt, batch_size, gamma, device)

        if steps_done % target_update_freq == 0:
            target.load_state_dict(policy.state_dict())

    rewards.append(total_reward)
    if len(rewards) > 10 and np.mean(rewards[-10:]) > 20:
        torch.save(policy.state_dict(), checkpoint_path)
        print(f'Modelo guardado - Recompensa media última 10: {np.mean(rewards[-10:]):.2f}')

  2%|▏         | 11/500 [01:04<55:18,  6.79s/it]

Modelo guardado - Recompensa media última 10: 145.00


  2%|▏         | 12/500 [01:09<51:28,  6.33s/it]

Modelo guardado - Recompensa media última 10: 140.50


  3%|▎         | 13/500 [01:15<49:31,  6.10s/it]

Modelo guardado - Recompensa media última 10: 150.50


  3%|▎         | 14/500 [01:24<56:21,  6.96s/it]

Modelo guardado - Recompensa media última 10: 167.00


  3%|▎         | 15/500 [01:31<56:07,  6.94s/it]

Modelo guardado - Recompensa media última 10: 193.00


  3%|▎         | 16/500 [01:36<52:23,  6.49s/it]

Modelo guardado - Recompensa media última 10: 187.50


  3%|▎         | 17/500 [01:41<47:48,  5.94s/it]

Modelo guardado - Recompensa media última 10: 189.00


  4%|▎         | 18/500 [01:46<46:00,  5.73s/it]

Modelo guardado - Recompensa media última 10: 194.00


  4%|▍         | 19/500 [01:58<1:00:18,  7.52s/it]

Modelo guardado - Recompensa media última 10: 212.00


  4%|▍         | 20/500 [02:05<58:22,  7.30s/it]  

Modelo guardado - Recompensa media última 10: 222.50


  4%|▍         | 21/500 [02:12<57:50,  7.25s/it]

Modelo guardado - Recompensa media última 10: 218.00


  4%|▍         | 22/500 [02:16<49:51,  6.26s/it]

Modelo guardado - Recompensa media última 10: 215.50


  5%|▍         | 23/500 [02:26<58:42,  7.39s/it]

Modelo guardado - Recompensa media última 10: 254.00


  5%|▍         | 24/500 [02:30<50:24,  6.35s/it]

Modelo guardado - Recompensa media última 10: 226.50


  5%|▌         | 25/500 [02:35<47:43,  6.03s/it]

Modelo guardado - Recompensa media última 10: 197.50


  5%|▌         | 26/500 [02:41<47:10,  5.97s/it]

Modelo guardado - Recompensa media última 10: 219.00


  5%|▌         | 27/500 [02:47<48:36,  6.17s/it]

Modelo guardado - Recompensa media última 10: 222.00


  6%|▌         | 28/500 [02:53<46:45,  5.94s/it]

Modelo guardado - Recompensa media última 10: 220.50


  6%|▌         | 29/500 [02:56<40:39,  5.18s/it]

Modelo guardado - Recompensa media última 10: 178.50


  6%|▌         | 30/500 [03:01<39:08,  5.00s/it]

Modelo guardado - Recompensa media última 10: 162.00


  6%|▌         | 31/500 [03:05<37:28,  4.79s/it]

Modelo guardado - Recompensa media última 10: 153.50


  6%|▋         | 32/500 [03:11<39:23,  5.05s/it]

Modelo guardado - Recompensa media última 10: 150.00


  7%|▋         | 33/500 [03:18<43:40,  5.61s/it]

Modelo guardado - Recompensa media última 10: 109.50


  7%|▋         | 34/500 [03:26<49:48,  6.41s/it]

Modelo guardado - Recompensa media última 10: 126.00


  7%|▋         | 35/500 [03:29<41:18,  5.33s/it]

Modelo guardado - Recompensa media última 10: 120.00


  7%|▋         | 36/500 [03:33<39:31,  5.11s/it]

Modelo guardado - Recompensa media última 10: 96.50


  7%|▋         | 37/500 [03:37<36:35,  4.74s/it]

Modelo guardado - Recompensa media última 10: 96.00


  8%|▊         | 38/500 [03:45<43:03,  5.59s/it]

Modelo guardado - Recompensa media última 10: 108.50


  8%|▊         | 39/500 [03:52<47:21,  6.16s/it]

Modelo guardado - Recompensa media última 10: 124.50


  8%|▊         | 40/500 [03:58<46:04,  6.01s/it]

Modelo guardado - Recompensa media última 10: 133.00


  8%|▊         | 41/500 [04:04<45:41,  5.97s/it]

Modelo guardado - Recompensa media última 10: 138.50


  8%|▊         | 42/500 [04:07<38:57,  5.10s/it]

Modelo guardado - Recompensa media última 10: 136.50


  9%|▊         | 43/500 [04:13<40:46,  5.35s/it]

Modelo guardado - Recompensa media última 10: 134.00


  9%|▉         | 44/500 [04:18<41:10,  5.42s/it]

Modelo guardado - Recompensa media última 10: 125.50


  9%|▉         | 45/500 [04:23<40:10,  5.30s/it]

Modelo guardado - Recompensa media última 10: 136.00


  9%|▉         | 46/500 [04:32<48:32,  6.42s/it]

Modelo guardado - Recompensa media última 10: 176.00


  9%|▉         | 47/500 [04:39<48:51,  6.47s/it]

Modelo guardado - Recompensa media última 10: 174.50


 10%|▉         | 48/500 [04:46<50:55,  6.76s/it]

Modelo guardado - Recompensa media última 10: 172.50


 10%|▉         | 49/500 [04:52<48:21,  6.43s/it]

Modelo guardado - Recompensa media última 10: 160.00


 10%|█         | 50/500 [05:00<51:00,  6.80s/it]

Modelo guardado - Recompensa media última 10: 168.00


 10%|█         | 51/500 [05:04<44:32,  5.95s/it]

Modelo guardado - Recompensa media última 10: 164.00


 10%|█         | 52/500 [05:11<48:05,  6.44s/it]

Modelo guardado - Recompensa media última 10: 185.00


 11%|█         | 53/500 [05:17<45:10,  6.06s/it]

Modelo guardado - Recompensa media última 10: 178.50


 11%|█         | 54/500 [05:26<52:56,  7.12s/it]

Modelo guardado - Recompensa media última 10: 198.00


 11%|█         | 55/500 [05:33<52:06,  7.03s/it]

Modelo guardado - Recompensa media última 10: 202.50


 11%|█         | 56/500 [05:44<1:00:19,  8.15s/it]

Modelo guardado - Recompensa media última 10: 176.50


 11%|█▏        | 57/500 [05:48<52:14,  7.08s/it]  

Modelo guardado - Recompensa media última 10: 183.00


 12%|█▏        | 58/500 [05:56<52:42,  7.15s/it]

Modelo guardado - Recompensa media última 10: 174.50


 12%|█▏        | 59/500 [05:59<44:05,  6.00s/it]

Modelo guardado - Recompensa media última 10: 170.00


 12%|█▏        | 60/500 [06:08<51:13,  6.99s/it]

Modelo guardado - Recompensa media última 10: 173.50


 12%|█▏        | 61/500 [06:12<44:35,  6.10s/it]

Modelo guardado - Recompensa media última 10: 174.50


 12%|█▏        | 62/500 [06:16<39:50,  5.46s/it]

Modelo guardado - Recompensa media última 10: 162.00


 13%|█▎        | 63/500 [06:21<37:51,  5.20s/it]

Modelo guardado - Recompensa media última 10: 163.50


 13%|█▎        | 64/500 [06:28<41:12,  5.67s/it]

Modelo guardado - Recompensa media última 10: 146.00


 13%|█▎        | 65/500 [06:34<41:55,  5.78s/it]

Modelo guardado - Recompensa media última 10: 142.00


 13%|█▎        | 66/500 [06:41<44:18,  6.13s/it]

Modelo guardado - Recompensa media última 10: 141.00


 13%|█▎        | 67/500 [06:45<39:39,  5.49s/it]

Modelo guardado - Recompensa media última 10: 136.50


 14%|█▎        | 68/500 [06:53<44:57,  6.24s/it]

Modelo guardado - Recompensa media última 10: 141.00


 14%|█▍        | 69/500 [06:59<46:06,  6.42s/it]

Modelo guardado - Recompensa media última 10: 158.00


 14%|█▍        | 70/500 [07:03<39:49,  5.56s/it]

Modelo guardado - Recompensa media última 10: 139.00


 14%|█▍        | 71/500 [07:06<33:31,  4.69s/it]

Modelo guardado - Recompensa media última 10: 129.00


 14%|█▍        | 72/500 [07:12<36:38,  5.14s/it]

Modelo guardado - Recompensa media última 10: 128.50


 15%|█▍        | 73/500 [07:20<42:22,  5.95s/it]

Modelo guardado - Recompensa media última 10: 161.00


 15%|█▍        | 74/500 [07:24<38:16,  5.39s/it]

Modelo guardado - Recompensa media última 10: 150.50


 15%|█▌        | 75/500 [07:31<42:44,  6.03s/it]

Modelo guardado - Recompensa media última 10: 157.00


 15%|█▌        | 76/500 [07:37<42:51,  6.06s/it]

Modelo guardado - Recompensa media última 10: 146.00


 15%|█▌        | 77/500 [07:43<42:09,  5.98s/it]

Modelo guardado - Recompensa media última 10: 145.00


 16%|█▌        | 78/500 [07:52<48:39,  6.92s/it]

Modelo guardado - Recompensa media última 10: 161.50


 16%|█▌        | 79/500 [08:00<50:09,  7.15s/it]

Modelo guardado - Recompensa media última 10: 173.50


 16%|█▌        | 80/500 [08:06<47:02,  6.72s/it]

Modelo guardado - Recompensa media última 10: 189.50


 16%|█▌        | 81/500 [08:12<45:25,  6.51s/it]

Modelo guardado - Recompensa media última 10: 201.00


 16%|█▋        | 82/500 [08:21<51:50,  7.44s/it]

Modelo guardado - Recompensa media última 10: 201.00


 17%|█▋        | 83/500 [08:29<52:47,  7.60s/it]

Modelo guardado - Recompensa media última 10: 178.50


 17%|█▋        | 84/500 [08:33<44:33,  6.43s/it]

Modelo guardado - Recompensa media última 10: 188.00


 17%|█▋        | 85/500 [08:45<56:26,  8.16s/it]

Modelo guardado - Recompensa media última 10: 213.50


 17%|█▋        | 86/500 [08:50<50:25,  7.31s/it]

Modelo guardado - Recompensa media última 10: 208.00


 17%|█▋        | 87/500 [08:55<44:58,  6.53s/it]

Modelo guardado - Recompensa media última 10: 205.50


 18%|█▊        | 88/500 [09:02<45:20,  6.60s/it]

Modelo guardado - Recompensa media última 10: 189.50


 18%|█▊        | 89/500 [09:06<38:53,  5.68s/it]

Modelo guardado - Recompensa media última 10: 163.00


 18%|█▊        | 90/500 [09:16<48:37,  7.12s/it]

Modelo guardado - Recompensa media última 10: 170.00


 18%|█▊        | 91/500 [09:24<49:55,  7.32s/it]

Modelo guardado - Recompensa media última 10: 179.00


 18%|█▊        | 92/500 [09:27<41:37,  6.12s/it]

Modelo guardado - Recompensa media última 10: 168.50


 19%|█▊        | 93/500 [09:35<45:15,  6.67s/it]

Modelo guardado - Recompensa media última 10: 171.00


 19%|█▉        | 94/500 [09:41<43:14,  6.39s/it]

Modelo guardado - Recompensa media última 10: 169.50


 19%|█▉        | 95/500 [09:47<43:33,  6.45s/it]

Modelo guardado - Recompensa media última 10: 134.50


 19%|█▉        | 96/500 [09:56<46:54,  6.97s/it]

Modelo guardado - Recompensa media última 10: 149.50


 19%|█▉        | 97/500 [09:59<39:41,  5.91s/it]

Modelo guardado - Recompensa media última 10: 150.50


 20%|█▉        | 98/500 [10:04<37:10,  5.55s/it]

Modelo guardado - Recompensa media última 10: 137.00


 20%|█▉        | 99/500 [10:12<43:16,  6.48s/it]

Modelo guardado - Recompensa media última 10: 154.00


 20%|██        | 100/500 [10:17<38:46,  5.82s/it]

Modelo guardado - Recompensa media última 10: 130.50


 20%|██        | 101/500 [10:25<44:00,  6.62s/it]

Modelo guardado - Recompensa media última 10: 129.00


 20%|██        | 102/500 [10:29<37:44,  5.69s/it]

Modelo guardado - Recompensa media última 10: 128.00


 21%|██        | 103/500 [10:33<34:46,  5.26s/it]

Modelo guardado - Recompensa media última 10: 109.00


 21%|██        | 104/500 [10:42<41:58,  6.36s/it]

Modelo guardado - Recompensa media última 10: 122.00


 21%|██        | 105/500 [10:52<49:00,  7.44s/it]

Modelo guardado - Recompensa media última 10: 145.50


 21%|██        | 106/500 [10:57<44:47,  6.82s/it]

Modelo guardado - Recompensa media última 10: 138.50


 21%|██▏       | 107/500 [11:04<45:37,  6.97s/it]

Modelo guardado - Recompensa media última 10: 153.50


 22%|██▏       | 108/500 [11:08<38:56,  5.96s/it]

Modelo guardado - Recompensa media última 10: 158.50


 22%|██▏       | 109/500 [11:14<38:15,  5.87s/it]

Modelo guardado - Recompensa media última 10: 153.00


 22%|██▏       | 110/500 [11:20<38:19,  5.90s/it]

Modelo guardado - Recompensa media última 10: 170.00


 22%|██▏       | 111/500 [11:27<40:20,  6.22s/it]

Modelo guardado - Recompensa media última 10: 164.00


 22%|██▏       | 112/500 [11:35<44:23,  6.86s/it]

Modelo guardado - Recompensa media última 10: 192.00


 23%|██▎       | 113/500 [11:47<53:15,  8.26s/it]

Modelo guardado - Recompensa media última 10: 234.00


 23%|██▎       | 114/500 [11:50<43:28,  6.76s/it]

Modelo guardado - Recompensa media última 10: 214.00


 23%|██▎       | 115/500 [11:55<40:36,  6.33s/it]

Modelo guardado - Recompensa media última 10: 184.00


 23%|██▎       | 116/500 [12:02<42:10,  6.59s/it]

Modelo guardado - Recompensa media última 10: 191.00


 23%|██▎       | 117/500 [12:11<45:22,  7.11s/it]

Modelo guardado - Recompensa media última 10: 195.50


 24%|██▎       | 118/500 [12:18<46:23,  7.29s/it]

Modelo guardado - Recompensa media última 10: 219.00


 24%|██▍       | 119/500 [12:28<50:14,  7.91s/it]

Modelo guardado - Recompensa media última 10: 244.00


 24%|██▍       | 120/500 [12:36<51:40,  8.16s/it]

Modelo guardado - Recompensa media última 10: 243.00


 24%|██▍       | 121/500 [12:42<47:11,  7.47s/it]

Modelo guardado - Recompensa media última 10: 251.00


 24%|██▍       | 122/500 [12:50<48:03,  7.63s/it]

Modelo guardado - Recompensa media última 10: 247.00


 25%|██▍       | 123/500 [12:56<43:54,  6.99s/it]

Modelo guardado - Recompensa media última 10: 217.00


 25%|██▍       | 124/500 [13:04<46:40,  7.45s/it]

Modelo guardado - Recompensa media última 10: 254.00


 25%|██▌       | 125/500 [13:10<43:08,  6.90s/it]

Modelo guardado - Recompensa media última 10: 269.50


 25%|██▌       | 126/500 [13:17<43:10,  6.93s/it]

Modelo guardado - Recompensa media última 10: 267.00


 25%|██▌       | 127/500 [13:23<41:08,  6.62s/it]

Modelo guardado - Recompensa media última 10: 245.00


 26%|██▌       | 128/500 [13:30<41:40,  6.72s/it]

Modelo guardado - Recompensa media última 10: 233.00


 26%|██▌       | 129/500 [13:33<35:32,  5.75s/it]

Modelo guardado - Recompensa media última 10: 196.50


 26%|██▌       | 130/500 [13:43<42:15,  6.85s/it]

Modelo guardado - Recompensa media última 10: 206.00


 26%|██▌       | 131/500 [13:48<38:36,  6.28s/it]

Modelo guardado - Recompensa media última 10: 193.50


 26%|██▋       | 132/500 [13:51<33:03,  5.39s/it]

Modelo guardado - Recompensa media última 10: 175.50


 27%|██▋       | 133/500 [13:59<38:15,  6.26s/it]

Modelo guardado - Recompensa media última 10: 177.00


 27%|██▋       | 134/500 [14:07<40:28,  6.63s/it]

Modelo guardado - Recompensa media última 10: 160.50


 27%|██▋       | 135/500 [14:15<42:57,  7.06s/it]

Modelo guardado - Recompensa media última 10: 155.50


 27%|██▋       | 136/500 [14:22<42:39,  7.03s/it]

Modelo guardado - Recompensa media última 10: 159.50


 27%|██▋       | 137/500 [14:32<47:36,  7.87s/it]

Modelo guardado - Recompensa media última 10: 188.00


 28%|██▊       | 138/500 [14:37<43:31,  7.21s/it]

Modelo guardado - Recompensa media última 10: 178.00


 28%|██▊       | 139/500 [14:44<41:41,  6.93s/it]

Modelo guardado - Recompensa media última 10: 183.00


 28%|██▊       | 140/500 [14:48<37:52,  6.31s/it]

Modelo guardado - Recompensa media última 10: 163.00


 28%|██▊       | 141/500 [14:57<41:14,  6.89s/it]

Modelo guardado - Recompensa media última 10: 175.00


 28%|██▊       | 142/500 [15:05<43:15,  7.25s/it]

Modelo guardado - Recompensa media última 10: 190.50


 29%|██▊       | 143/500 [15:11<40:50,  6.86s/it]

Modelo guardado - Recompensa media última 10: 188.50


 29%|██▉       | 144/500 [15:16<37:16,  6.28s/it]

Modelo guardado - Recompensa media última 10: 175.50


 29%|██▉       | 145/500 [15:23<38:27,  6.50s/it]

Modelo guardado - Recompensa media última 10: 177.00


 29%|██▉       | 146/500 [15:26<32:44,  5.55s/it]

Modelo guardado - Recompensa media última 10: 159.50


 29%|██▉       | 147/500 [15:34<37:26,  6.36s/it]

Modelo guardado - Recompensa media última 10: 147.50


 30%|██▉       | 148/500 [15:41<37:58,  6.47s/it]

Modelo guardado - Recompensa media última 10: 150.00


 30%|██▉       | 149/500 [15:46<36:09,  6.18s/it]

Modelo guardado - Recompensa media última 10: 154.00


 30%|███       | 150/500 [15:52<35:12,  6.04s/it]

Modelo guardado - Recompensa media última 10: 162.00


 30%|███       | 151/500 [15:58<34:38,  5.95s/it]

Modelo guardado - Recompensa media última 10: 154.00


 30%|███       | 152/500 [16:05<36:23,  6.28s/it]

Modelo guardado - Recompensa media última 10: 164.00


 31%|███       | 153/500 [16:11<36:38,  6.34s/it]

Modelo guardado - Recompensa media última 10: 164.50


 31%|███       | 154/500 [16:17<35:00,  6.07s/it]

Modelo guardado - Recompensa media última 10: 160.00


 31%|███       | 155/500 [16:23<34:53,  6.07s/it]

Modelo guardado - Recompensa media última 10: 161.50


 31%|███       | 156/500 [16:33<42:26,  7.40s/it]

Modelo guardado - Recompensa media última 10: 205.00


 31%|███▏      | 157/500 [16:40<40:10,  7.03s/it]

Modelo guardado - Recompensa media última 10: 200.50


 32%|███▏      | 158/500 [16:50<45:48,  8.04s/it]

Modelo guardado - Recompensa media última 10: 221.00


 32%|███▏      | 159/500 [16:54<38:52,  6.84s/it]

Modelo guardado - Recompensa media última 10: 215.50


 32%|███▏      | 160/500 [16:58<33:04,  5.84s/it]

Modelo guardado - Recompensa media última 10: 205.00


 32%|███▏      | 161/500 [17:03<31:57,  5.66s/it]

Modelo guardado - Recompensa media última 10: 202.50


 32%|███▏      | 162/500 [17:13<40:17,  7.15s/it]

Modelo guardado - Recompensa media última 10: 226.50


 33%|███▎      | 163/500 [17:23<43:48,  7.80s/it]

Modelo guardado - Recompensa media última 10: 239.00


 33%|███▎      | 164/500 [17:32<45:28,  8.12s/it]

Modelo guardado - Recompensa media última 10: 258.00


 33%|███▎      | 165/500 [17:36<38:48,  6.95s/it]

Modelo guardado - Recompensa media última 10: 245.00


 33%|███▎      | 166/500 [17:40<34:18,  6.16s/it]

Modelo guardado - Recompensa media última 10: 203.50


 33%|███▎      | 167/500 [17:46<34:14,  6.17s/it]

Modelo guardado - Recompensa media última 10: 191.50


 34%|███▎      | 168/500 [17:56<39:51,  7.20s/it]

Modelo guardado - Recompensa media última 10: 187.00


 34%|███▍      | 169/500 [18:02<37:07,  6.73s/it]

Modelo guardado - Recompensa media última 10: 195.00


 34%|███▍      | 170/500 [18:10<39:38,  7.21s/it]

Modelo guardado - Recompensa media última 10: 221.00


 34%|███▍      | 171/500 [18:18<41:41,  7.60s/it]

Modelo guardado - Recompensa media última 10: 231.50


 34%|███▍      | 172/500 [18:23<36:35,  6.69s/it]

Modelo guardado - Recompensa media última 10: 185.50


 35%|███▍      | 173/500 [18:29<34:46,  6.38s/it]

Modelo guardado - Recompensa media última 10: 178.00


 35%|███▍      | 174/500 [18:35<34:28,  6.34s/it]

Modelo guardado - Recompensa media última 10: 167.50


 35%|███▌      | 175/500 [18:39<31:05,  5.74s/it]

Modelo guardado - Recompensa media última 10: 177.00


 35%|███▌      | 176/500 [18:44<29:27,  5.46s/it]

Modelo guardado - Recompensa media última 10: 173.00


 35%|███▌      | 177/500 [18:48<26:11,  4.86s/it]

Modelo guardado - Recompensa media última 10: 170.50


 36%|███▌      | 178/500 [18:56<31:26,  5.86s/it]

Modelo guardado - Recompensa media última 10: 173.50


 36%|███▌      | 179/500 [19:04<35:23,  6.62s/it]

Modelo guardado - Recompensa media última 10: 190.00


 36%|███▌      | 180/500 [19:07<29:46,  5.58s/it]

Modelo guardado - Recompensa media última 10: 159.00


 36%|███▌      | 181/500 [19:15<32:51,  6.18s/it]

Modelo guardado - Recompensa media última 10: 163.00


 36%|███▋      | 182/500 [19:20<31:52,  6.01s/it]

Modelo guardado - Recompensa media última 10: 161.50


 37%|███▋      | 183/500 [19:26<30:18,  5.74s/it]

Modelo guardado - Recompensa media última 10: 161.50


 37%|███▋      | 184/500 [19:31<30:18,  5.75s/it]

Modelo guardado - Recompensa media última 10: 152.00


 37%|███▋      | 185/500 [19:39<33:11,  6.32s/it]

Modelo guardado - Recompensa media última 10: 145.50


 37%|███▋      | 186/500 [19:49<38:23,  7.34s/it]

Modelo guardado - Recompensa media última 10: 175.50


 37%|███▋      | 187/500 [19:54<35:29,  6.80s/it]

Modelo guardado - Recompensa media última 10: 186.50


 38%|███▊      | 188/500 [20:01<35:38,  6.85s/it]

Modelo guardado - Recompensa media última 10: 175.00


 38%|███▊      | 189/500 [20:06<32:30,  6.27s/it]

Modelo guardado - Recompensa media última 10: 153.00


 38%|███▊      | 190/500 [20:15<36:38,  7.09s/it]

Modelo guardado - Recompensa media última 10: 179.00


 38%|███▊      | 191/500 [20:21<34:36,  6.72s/it]

Modelo guardado - Recompensa media última 10: 161.50


 38%|███▊      | 192/500 [20:27<32:38,  6.36s/it]

Modelo guardado - Recompensa media última 10: 167.00


 39%|███▊      | 193/500 [20:32<30:41,  6.00s/it]

Modelo guardado - Recompensa media última 10: 164.00


 39%|███▉      | 194/500 [20:43<39:06,  7.67s/it]

Modelo guardado - Recompensa media última 10: 185.00


 39%|███▉      | 195/500 [20:47<32:27,  6.39s/it]

Modelo guardado - Recompensa media última 10: 187.50


 39%|███▉      | 196/500 [20:52<31:20,  6.19s/it]

Modelo guardado - Recompensa media última 10: 175.00


 39%|███▉      | 197/500 [20:57<28:41,  5.68s/it]

Modelo guardado - Recompensa media última 10: 165.50


 40%|███▉      | 198/500 [21:02<28:26,  5.65s/it]

Modelo guardado - Recompensa media última 10: 155.00


 40%|███▉      | 199/500 [21:08<27:44,  5.53s/it]

Modelo guardado - Recompensa media última 10: 147.00


 40%|████      | 200/500 [21:17<33:38,  6.73s/it]

Modelo guardado - Recompensa media última 10: 141.00


 40%|████      | 201/500 [21:29<41:36,  8.35s/it]

Modelo guardado - Recompensa media última 10: 214.00


 40%|████      | 202/500 [21:35<37:42,  7.59s/it]

Modelo guardado - Recompensa media última 10: 220.50


 41%|████      | 203/500 [21:42<36:03,  7.29s/it]

Modelo guardado - Recompensa media última 10: 217.50


 41%|████      | 204/500 [21:49<36:13,  7.34s/it]

Modelo guardado - Recompensa media última 10: 207.50


 41%|████      | 205/500 [21:57<36:11,  7.36s/it]

Modelo guardado - Recompensa media última 10: 222.00


 41%|████      | 206/500 [22:05<38:01,  7.76s/it]

Modelo guardado - Recompensa media última 10: 233.50


 41%|████▏     | 207/500 [22:14<38:43,  7.93s/it]

Modelo guardado - Recompensa media última 10: 270.50


 42%|████▏     | 208/500 [22:19<34:59,  7.19s/it]

Modelo guardado - Recompensa media última 10: 279.00


 42%|████▏     | 209/500 [22:23<30:08,  6.21s/it]

Modelo guardado - Recompensa media última 10: 282.50


 42%|████▏     | 210/500 [22:29<29:08,  6.03s/it]

Modelo guardado - Recompensa media última 10: 263.00


 42%|████▏     | 211/500 [22:34<28:23,  5.90s/it]

Modelo guardado - Recompensa media última 10: 196.50


 42%|████▏     | 212/500 [22:43<32:43,  6.82s/it]

Modelo guardado - Recompensa media última 10: 195.50


 43%|████▎     | 213/500 [22:49<31:28,  6.58s/it]

Modelo guardado - Recompensa media última 10: 197.50


 43%|████▎     | 214/500 [22:56<32:00,  6.72s/it]

Modelo guardado - Recompensa media última 10: 195.50


 43%|████▎     | 215/500 [23:01<29:07,  6.13s/it]

Modelo guardado - Recompensa media última 10: 175.00


 43%|████▎     | 216/500 [23:09<31:31,  6.66s/it]

Modelo guardado - Recompensa media última 10: 164.00


 43%|████▎     | 217/500 [23:15<30:48,  6.53s/it]

Modelo guardado - Recompensa media última 10: 134.00


 44%|████▎     | 218/500 [23:23<32:43,  6.96s/it]

Modelo guardado - Recompensa media última 10: 134.50


 44%|████▍     | 219/500 [23:30<32:31,  6.95s/it]

Modelo guardado - Recompensa media última 10: 149.50


 44%|████▍     | 220/500 [23:47<46:30,  9.97s/it]

Modelo guardado - Recompensa media última 10: 201.50


 44%|████▍     | 221/500 [23:56<44:21,  9.54s/it]

Modelo guardado - Recompensa media última 10: 215.00


 44%|████▍     | 222/500 [24:02<39:30,  8.53s/it]

Modelo guardado - Recompensa media última 10: 205.00


 45%|████▍     | 223/500 [24:08<36:02,  7.81s/it]

Modelo guardado - Recompensa media última 10: 219.00


 45%|████▍     | 224/500 [24:15<35:18,  7.68s/it]

Modelo guardado - Recompensa media última 10: 221.50


 45%|████▌     | 225/500 [24:19<30:28,  6.65s/it]

Modelo guardado - Recompensa media última 10: 224.50


 45%|████▌     | 226/500 [24:28<32:16,  7.07s/it]

Modelo guardado - Recompensa media última 10: 253.00


 45%|████▌     | 227/500 [24:34<30:46,  6.76s/it]

Modelo guardado - Recompensa media última 10: 249.00


 46%|████▌     | 228/500 [24:38<27:28,  6.06s/it]

Modelo guardado - Recompensa media última 10: 240.50


 46%|████▌     | 229/500 [24:47<31:33,  6.99s/it]

Modelo guardado - Recompensa media última 10: 246.00


 46%|████▌     | 230/500 [24:53<29:41,  6.60s/it]

Modelo guardado - Recompensa media última 10: 202.00


 46%|████▌     | 231/500 [24:58<28:16,  6.31s/it]

Modelo guardado - Recompensa media última 10: 185.50


 46%|████▋     | 232/500 [25:07<30:41,  6.87s/it]

Modelo guardado - Recompensa media última 10: 204.00


 47%|████▋     | 233/500 [25:12<28:35,  6.43s/it]

Modelo guardado - Recompensa media última 10: 187.00


 47%|████▋     | 234/500 [25:20<30:49,  6.95s/it]

Modelo guardado - Recompensa media última 10: 211.00


 47%|████▋     | 235/500 [25:29<32:45,  7.42s/it]

Modelo guardado - Recompensa media última 10: 217.00


 47%|████▋     | 236/500 [25:37<33:14,  7.56s/it]

Modelo guardado - Recompensa media última 10: 179.00


 47%|████▋     | 237/500 [25:49<40:02,  9.13s/it]

Modelo guardado - Recompensa media última 10: 201.50


 48%|████▊     | 238/500 [25:57<37:27,  8.58s/it]

Modelo guardado - Recompensa media última 10: 211.50


 48%|████▊     | 239/500 [26:00<30:22,  6.98s/it]

Modelo guardado - Recompensa media última 10: 186.00


 48%|████▊     | 240/500 [26:13<37:35,  8.67s/it]

Modelo guardado - Recompensa media última 10: 226.50


 48%|████▊     | 241/500 [26:18<32:38,  7.56s/it]

Modelo guardado - Recompensa media última 10: 224.50


 48%|████▊     | 242/500 [26:23<30:05,  7.00s/it]

Modelo guardado - Recompensa media última 10: 205.50


 49%|████▊     | 243/500 [26:29<28:28,  6.65s/it]

Modelo guardado - Recompensa media última 10: 212.00


 49%|████▉     | 244/500 [26:36<28:58,  6.79s/it]

Modelo guardado - Recompensa media última 10: 194.00


 49%|████▉     | 245/500 [26:42<27:57,  6.58s/it]

Modelo guardado - Recompensa media última 10: 191.00


 49%|████▉     | 246/500 [26:49<28:10,  6.66s/it]

Modelo guardado - Recompensa media última 10: 195.50


 49%|████▉     | 247/500 [26:57<29:56,  7.10s/it]

Modelo guardado - Recompensa media última 10: 172.50


 50%|████▉     | 248/500 [27:10<36:27,  8.68s/it]

Modelo guardado - Recompensa media última 10: 177.50


 50%|████▉     | 249/500 [27:15<32:23,  7.74s/it]

Modelo guardado - Recompensa media última 10: 192.50


 50%|█████     | 250/500 [27:25<34:51,  8.37s/it]

Modelo guardado - Recompensa media última 10: 170.50


 50%|█████     | 251/500 [27:28<28:27,  6.86s/it]

Modelo guardado - Recompensa media última 10: 172.00


 50%|█████     | 252/500 [27:37<30:00,  7.26s/it]

Modelo guardado - Recompensa media última 10: 186.00


 51%|█████     | 253/500 [27:45<31:11,  7.58s/it]

Modelo guardado - Recompensa media última 10: 210.00


 51%|█████     | 254/500 [28:02<42:28, 10.36s/it]

Modelo guardado - Recompensa media última 10: 251.00


 51%|█████     | 255/500 [28:08<37:36,  9.21s/it]

Modelo guardado - Recompensa media última 10: 247.50


 51%|█████     | 256/500 [28:15<34:45,  8.55s/it]

Modelo guardado - Recompensa media última 10: 248.00


 51%|█████▏    | 257/500 [28:23<33:14,  8.21s/it]

Modelo guardado - Recompensa media última 10: 261.50


 52%|█████▏    | 258/500 [28:28<29:04,  7.21s/it]

Modelo guardado - Recompensa media última 10: 239.00


 52%|█████▏    | 259/500 [28:36<30:19,  7.55s/it]

Modelo guardado - Recompensa media última 10: 238.50


 52%|█████▏    | 260/500 [28:39<25:05,  6.27s/it]

Modelo guardado - Recompensa media última 10: 217.00


 52%|█████▏    | 261/500 [28:47<26:31,  6.66s/it]

Modelo guardado - Recompensa media última 10: 244.50


 52%|█████▏    | 262/500 [28:55<28:30,  7.19s/it]

Modelo guardado - Recompensa media última 10: 232.50


 53%|█████▎    | 263/500 [28:59<23:51,  6.04s/it]

Modelo guardado - Recompensa media última 10: 194.50


 53%|█████▎    | 264/500 [29:05<23:48,  6.05s/it]

Modelo guardado - Recompensa media última 10: 140.50


 53%|█████▎    | 265/500 [29:08<20:18,  5.18s/it]

Modelo guardado - Recompensa media última 10: 141.00


 53%|█████▎    | 266/500 [29:18<25:36,  6.57s/it]

Modelo guardado - Recompensa media última 10: 150.00


 53%|█████▎    | 267/500 [29:23<24:26,  6.29s/it]

Modelo guardado - Recompensa media última 10: 146.00


 54%|█████▎    | 268/500 [29:32<27:02,  6.99s/it]

Modelo guardado - Recompensa media última 10: 181.00


 54%|█████▍    | 269/500 [29:38<25:32,  6.64s/it]

Modelo guardado - Recompensa media última 10: 181.00


 54%|█████▍    | 270/500 [29:48<29:11,  7.62s/it]

Modelo guardado - Recompensa media última 10: 224.00


 54%|█████▍    | 271/500 [30:01<35:15,  9.24s/it]

Modelo guardado - Recompensa media última 10: 215.00


 54%|█████▍    | 272/500 [30:06<30:58,  8.15s/it]

Modelo guardado - Recompensa media última 10: 219.50


 55%|█████▍    | 273/500 [30:12<28:32,  7.54s/it]

Modelo guardado - Recompensa media última 10: 227.50


 55%|█████▍    | 274/500 [30:24<33:23,  8.86s/it]

Modelo guardado - Recompensa media última 10: 272.50


 55%|█████▌    | 275/500 [30:30<29:50,  7.96s/it]

Modelo guardado - Recompensa media última 10: 278.50


 55%|█████▌    | 276/500 [30:36<27:32,  7.38s/it]

Modelo guardado - Recompensa media última 10: 273.00


 55%|█████▌    | 277/500 [30:45<28:35,  7.69s/it]

Modelo guardado - Recompensa media última 10: 292.00


 56%|█████▌    | 278/500 [30:51<26:34,  7.18s/it]

Modelo guardado - Recompensa media última 10: 279.00


 56%|█████▌    | 279/500 [31:01<29:41,  8.06s/it]

Modelo guardado - Recompensa media última 10: 305.50


 56%|█████▌    | 280/500 [31:09<30:15,  8.25s/it]

Modelo guardado - Recompensa media última 10: 274.00


 56%|█████▌    | 281/500 [31:15<27:14,  7.46s/it]

Modelo guardado - Recompensa media última 10: 264.50


 56%|█████▋    | 282/500 [31:19<23:03,  6.35s/it]

Modelo guardado - Recompensa media última 10: 252.50


 57%|█████▋    | 283/500 [31:28<26:40,  7.38s/it]

Modelo guardado - Recompensa media última 10: 263.50


 57%|█████▋    | 284/500 [31:35<26:04,  7.24s/it]

Modelo guardado - Recompensa media última 10: 228.00


 57%|█████▋    | 285/500 [31:45<27:57,  7.80s/it]

Modelo guardado - Recompensa media última 10: 243.50


 57%|█████▋    | 286/500 [31:49<24:35,  6.89s/it]

Modelo guardado - Recompensa media última 10: 227.50


 57%|█████▋    | 287/500 [31:59<27:29,  7.74s/it]

Modelo guardado - Recompensa media última 10: 249.00


 58%|█████▊    | 288/500 [32:05<25:01,  7.08s/it]

Modelo guardado - Recompensa media última 10: 231.00


 58%|█████▊    | 289/500 [32:11<23:42,  6.74s/it]

Modelo guardado - Recompensa media última 10: 198.00


 58%|█████▊    | 290/500 [32:18<24:00,  6.86s/it]

Modelo guardado - Recompensa media última 10: 199.00


 58%|█████▊    | 291/500 [32:24<22:51,  6.56s/it]

Modelo guardado - Recompensa media última 10: 195.00


 58%|█████▊    | 292/500 [32:32<24:13,  6.99s/it]

Modelo guardado - Recompensa media última 10: 215.50


 59%|█████▊    | 293/500 [32:41<27:12,  7.89s/it]

Modelo guardado - Recompensa media última 10: 224.50


 59%|█████▉    | 294/500 [32:51<28:41,  8.36s/it]

Modelo guardado - Recompensa media última 10: 233.00


 59%|█████▉    | 295/500 [32:56<25:28,  7.45s/it]

Modelo guardado - Recompensa media última 10: 215.00


 59%|█████▉    | 296/500 [33:02<24:03,  7.07s/it]

Modelo guardado - Recompensa media última 10: 231.00


 59%|█████▉    | 297/500 [33:08<22:44,  6.72s/it]

Modelo guardado - Recompensa media última 10: 187.50


 60%|█████▉    | 298/500 [33:13<20:06,  5.97s/it]

Modelo guardado - Recompensa media última 10: 184.00


 60%|█████▉    | 299/500 [33:21<22:10,  6.62s/it]

Modelo guardado - Recompensa media última 10: 195.00


 60%|██████    | 300/500 [33:32<27:10,  8.15s/it]

Modelo guardado - Recompensa media última 10: 209.00


 60%|██████    | 301/500 [33:38<24:01,  7.25s/it]

Modelo guardado - Recompensa media última 10: 204.00


 60%|██████    | 302/500 [33:45<23:44,  7.19s/it]

Modelo guardado - Recompensa media última 10: 195.00


 61%|██████    | 303/500 [33:50<22:15,  6.78s/it]

Modelo guardado - Recompensa media última 10: 173.50


 61%|██████    | 304/500 [34:07<31:57,  9.78s/it]

Modelo guardado - Recompensa media última 10: 190.00


 61%|██████    | 305/500 [34:12<27:13,  8.38s/it]

Modelo guardado - Recompensa media última 10: 184.50


 61%|██████    | 306/500 [34:18<24:17,  7.51s/it]

Modelo guardado - Recompensa media última 10: 173.50


 61%|██████▏   | 307/500 [34:32<30:39,  9.53s/it]

Modelo guardado - Recompensa media última 10: 199.50


 62%|██████▏   | 308/500 [34:37<25:41,  8.03s/it]

Modelo guardado - Recompensa media última 10: 200.00


 62%|██████▏   | 309/500 [34:42<22:59,  7.22s/it]

Modelo guardado - Recompensa media última 10: 185.50


 62%|██████▏   | 310/500 [34:50<23:49,  7.53s/it]

Modelo guardado - Recompensa media última 10: 164.50


 62%|██████▏   | 311/500 [34:58<23:31,  7.47s/it]

Modelo guardado - Recompensa media última 10: 178.00


 62%|██████▏   | 312/500 [35:07<25:19,  8.08s/it]

Modelo guardado - Recompensa media última 10: 203.50


 63%|██████▎   | 313/500 [35:16<25:59,  8.34s/it]

Modelo guardado - Recompensa media última 10: 211.00


 63%|██████▎   | 314/500 [35:21<22:28,  7.25s/it]

Modelo guardado - Recompensa media última 10: 183.00


 63%|██████▎   | 315/500 [35:24<18:54,  6.13s/it]

Modelo guardado - Recompensa media última 10: 179.00


 63%|██████▎   | 316/500 [35:31<19:28,  6.35s/it]

Modelo guardado - Recompensa media última 10: 179.50


 63%|██████▎   | 317/500 [35:41<22:58,  7.53s/it]

Modelo guardado - Recompensa media última 10: 166.00


 64%|██████▎   | 318/500 [35:45<18:57,  6.25s/it]

Modelo guardado - Recompensa media última 10: 165.00


 64%|██████▍   | 319/500 [35:51<18:39,  6.19s/it]

Modelo guardado - Recompensa media última 10: 172.00


 64%|██████▍   | 320/500 [35:58<19:10,  6.39s/it]

Modelo guardado - Recompensa media última 10: 171.50


 64%|██████▍   | 321/500 [36:01<16:10,  5.42s/it]

Modelo guardado - Recompensa media última 10: 158.00


 64%|██████▍   | 322/500 [36:05<15:23,  5.19s/it]

Modelo guardado - Recompensa media última 10: 124.00


 65%|██████▍   | 323/500 [36:14<18:29,  6.27s/it]

Modelo guardado - Recompensa media última 10: 129.00


 65%|██████▍   | 324/500 [36:23<20:27,  6.98s/it]

Modelo guardado - Recompensa media última 10: 125.50


 65%|██████▌   | 325/500 [36:29<20:04,  6.88s/it]

Modelo guardado - Recompensa media última 10: 138.50


 65%|██████▌   | 326/500 [36:35<19:02,  6.57s/it]

Modelo guardado - Recompensa media última 10: 139.50


 65%|██████▌   | 327/500 [36:40<17:26,  6.05s/it]

Modelo guardado - Recompensa media última 10: 126.50


 66%|██████▌   | 328/500 [36:48<19:04,  6.66s/it]

Modelo guardado - Recompensa media última 10: 144.50


 66%|██████▌   | 329/500 [36:56<20:07,  7.06s/it]

Modelo guardado - Recompensa media última 10: 154.50


 66%|██████▌   | 330/500 [37:03<19:36,  6.92s/it]

Modelo guardado - Recompensa media última 10: 151.00


 66%|██████▌   | 331/500 [37:10<19:51,  7.05s/it]

Modelo guardado - Recompensa media última 10: 155.50


 66%|██████▋   | 332/500 [37:15<18:11,  6.50s/it]

Modelo guardado - Recompensa media última 10: 158.00


 67%|██████▋   | 333/500 [37:24<19:48,  7.12s/it]

Modelo guardado - Recompensa media última 10: 175.00


 67%|██████▋   | 334/500 [37:30<18:30,  6.69s/it]

Modelo guardado - Recompensa media última 10: 171.00


 67%|██████▋   | 335/500 [37:39<20:20,  7.40s/it]

Modelo guardado - Recompensa media última 10: 181.00


 67%|██████▋   | 336/500 [37:47<21:17,  7.79s/it]

Modelo guardado - Recompensa media última 10: 206.50


 67%|██████▋   | 337/500 [37:57<22:36,  8.32s/it]

Modelo guardado - Recompensa media última 10: 214.50


 68%|██████▊   | 338/500 [38:04<21:34,  7.99s/it]

Modelo guardado - Recompensa media última 10: 214.00


 68%|██████▊   | 339/500 [38:08<17:47,  6.63s/it]

Modelo guardado - Recompensa media última 10: 192.50


 68%|██████▊   | 340/500 [38:13<17:00,  6.38s/it]

Modelo guardado - Recompensa media última 10: 194.00


 68%|██████▊   | 341/500 [38:19<16:43,  6.31s/it]

Modelo guardado - Recompensa media última 10: 195.00


 68%|██████▊   | 342/500 [38:25<16:18,  6.19s/it]

Modelo guardado - Recompensa media última 10: 201.50


 69%|██████▊   | 343/500 [38:37<20:05,  7.68s/it]

Modelo guardado - Recompensa media última 10: 213.50


 69%|██████▉   | 344/500 [38:45<20:41,  7.96s/it]

Modelo guardado - Recompensa media última 10: 218.50


 69%|██████▉   | 345/500 [38:51<18:55,  7.33s/it]

Modelo guardado - Recompensa media última 10: 206.50


 69%|██████▉   | 346/500 [38:59<19:32,  7.62s/it]

Modelo guardado - Recompensa media última 10: 217.50


 69%|██████▉   | 347/500 [39:07<19:30,  7.65s/it]

Modelo guardado - Recompensa media última 10: 209.00


 70%|██████▉   | 348/500 [39:19<22:52,  9.03s/it]

Modelo guardado - Recompensa media última 10: 211.50


 70%|██████▉   | 349/500 [39:28<22:42,  9.02s/it]

Modelo guardado - Recompensa media última 10: 234.50


 70%|███████   | 350/500 [39:37<21:59,  8.80s/it]

Modelo guardado - Recompensa media última 10: 249.00


 70%|███████   | 351/500 [39:43<20:27,  8.24s/it]

Modelo guardado - Recompensa media última 10: 250.00


 70%|███████   | 352/500 [39:51<19:27,  7.89s/it]

Modelo guardado - Recompensa media última 10: 248.50


 71%|███████   | 353/500 [40:02<21:52,  8.93s/it]

Modelo guardado - Recompensa media última 10: 222.00


 71%|███████   | 354/500 [40:08<19:22,  7.96s/it]

Modelo guardado - Recompensa media última 10: 218.00


 71%|███████   | 355/500 [40:15<18:33,  7.68s/it]

Modelo guardado - Recompensa media última 10: 225.50


 71%|███████   | 356/500 [40:19<16:10,  6.74s/it]

Modelo guardado - Recompensa media última 10: 187.50


 71%|███████▏  | 357/500 [40:26<16:11,  6.80s/it]

Modelo guardado - Recompensa media última 10: 188.50


 72%|███████▏  | 358/500 [40:34<16:30,  6.98s/it]

Modelo guardado - Recompensa media última 10: 185.00


 72%|███████▏  | 359/500 [40:41<16:45,  7.13s/it]

Modelo guardado - Recompensa media última 10: 180.50


 72%|███████▏  | 360/500 [40:53<19:42,  8.45s/it]

Modelo guardado - Recompensa media última 10: 175.00


 72%|███████▏  | 361/500 [41:00<18:51,  8.14s/it]

Modelo guardado - Recompensa media última 10: 173.50


 72%|███████▏  | 362/500 [41:05<16:55,  7.36s/it]

Modelo guardado - Recompensa media última 10: 173.50


 73%|███████▎  | 363/500 [41:10<15:09,  6.64s/it]

Modelo guardado - Recompensa media última 10: 154.50


 73%|███████▎  | 364/500 [41:18<15:56,  7.04s/it]

Modelo guardado - Recompensa media última 10: 162.00


 73%|███████▎  | 365/500 [41:22<13:10,  5.86s/it]

Modelo guardado - Recompensa media última 10: 149.50


 73%|███████▎  | 366/500 [41:31<15:22,  6.89s/it]

Modelo guardado - Recompensa media última 10: 172.50


 73%|███████▎  | 367/500 [41:38<15:40,  7.07s/it]

Modelo guardado - Recompensa media última 10: 177.50


 74%|███████▎  | 368/500 [41:45<15:05,  6.86s/it]

Modelo guardado - Recompensa media última 10: 168.00


 74%|███████▍  | 369/500 [41:54<16:34,  7.59s/it]

Modelo guardado - Recompensa media última 10: 176.00


 74%|███████▍  | 370/500 [42:03<17:19,  7.99s/it]

Modelo guardado - Recompensa media última 10: 186.50


 74%|███████▍  | 371/500 [42:07<14:37,  6.80s/it]

Modelo guardado - Recompensa media última 10: 182.50


 74%|███████▍  | 372/500 [42:13<13:51,  6.50s/it]

Modelo guardado - Recompensa media última 10: 182.50


 75%|███████▍  | 373/500 [42:17<12:20,  5.83s/it]

Modelo guardado - Recompensa media última 10: 186.00


 75%|███████▍  | 374/500 [42:24<13:04,  6.23s/it]

Modelo guardado - Recompensa media última 10: 195.50


 75%|███████▌  | 375/500 [42:32<14:03,  6.74s/it]

Modelo guardado - Recompensa media última 10: 200.00


 75%|███████▌  | 376/500 [42:40<14:36,  7.07s/it]

Modelo guardado - Recompensa media última 10: 186.00


 75%|███████▌  | 377/500 [42:47<14:39,  7.15s/it]

Modelo guardado - Recompensa media última 10: 174.50


 76%|███████▌  | 378/500 [42:54<14:20,  7.05s/it]

Modelo guardado - Recompensa media última 10: 187.00


 76%|███████▌  | 379/500 [43:01<14:06,  7.00s/it]

Modelo guardado - Recompensa media última 10: 195.50


 76%|███████▌  | 380/500 [43:07<13:20,  6.67s/it]

Modelo guardado - Recompensa media última 10: 176.50


 76%|███████▌  | 381/500 [43:18<15:35,  7.86s/it]

Modelo guardado - Recompensa media última 10: 208.50


 76%|███████▋  | 382/500 [43:26<15:51,  8.06s/it]

Modelo guardado - Recompensa media última 10: 216.50


 77%|███████▋  | 383/500 [43:35<16:30,  8.46s/it]

Modelo guardado - Recompensa media última 10: 249.50


 77%|███████▋  | 384/500 [43:40<14:12,  7.35s/it]

Modelo guardado - Recompensa media última 10: 231.00


 77%|███████▋  | 385/500 [43:51<16:00,  8.35s/it]

Modelo guardado - Recompensa media última 10: 274.50


 77%|███████▋  | 386/500 [44:06<19:33, 10.30s/it]

Modelo guardado - Recompensa media última 10: 288.50


 77%|███████▋  | 387/500 [44:11<16:29,  8.76s/it]

Modelo guardado - Recompensa media última 10: 287.00


 78%|███████▊  | 388/500 [44:18<15:41,  8.41s/it]

Modelo guardado - Recompensa media última 10: 292.00


 78%|███████▊  | 389/500 [44:27<15:31,  8.39s/it]

Modelo guardado - Recompensa media última 10: 270.00


 78%|███████▊  | 390/500 [44:36<15:34,  8.50s/it]

Modelo guardado - Recompensa media última 10: 299.00


 78%|███████▊  | 391/500 [44:43<14:45,  8.13s/it]

Modelo guardado - Recompensa media última 10: 285.50


 78%|███████▊  | 392/500 [44:47<12:32,  6.97s/it]

Modelo guardado - Recompensa media última 10: 267.50


 79%|███████▊  | 393/500 [44:54<12:31,  7.03s/it]

Modelo guardado - Recompensa media última 10: 241.00


 79%|███████▉  | 394/500 [45:02<12:54,  7.31s/it]

Modelo guardado - Recompensa media última 10: 249.50


 79%|███████▉  | 395/500 [45:11<13:24,  7.66s/it]

Modelo guardado - Recompensa media última 10: 244.00


 79%|███████▉  | 396/500 [45:18<13:19,  7.69s/it]

Modelo guardado - Recompensa media última 10: 234.50


 79%|███████▉  | 397/500 [45:24<12:15,  7.14s/it]

Modelo guardado - Recompensa media última 10: 243.00


 80%|███████▉  | 398/500 [45:35<13:53,  8.17s/it]

Modelo guardado - Recompensa media última 10: 245.00


 80%|███████▉  | 399/500 [45:43<13:34,  8.06s/it]

Modelo guardado - Recompensa media última 10: 252.00


 80%|████████  | 400/500 [45:54<15:05,  9.06s/it]

Modelo guardado - Recompensa media última 10: 235.50


 80%|████████  | 401/500 [46:07<17:00, 10.31s/it]

Modelo guardado - Recompensa media última 10: 253.00


 80%|████████  | 402/500 [46:14<15:17,  9.36s/it]

Modelo guardado - Recompensa media última 10: 268.50


 81%|████████  | 403/500 [46:21<13:31,  8.37s/it]

Modelo guardado - Recompensa media última 10: 260.50


 81%|████████  | 404/500 [46:27<12:36,  7.88s/it]

Modelo guardado - Recompensa media última 10: 257.50


 81%|████████  | 405/500 [46:33<11:33,  7.30s/it]

Modelo guardado - Recompensa media última 10: 218.50


 81%|████████  | 406/500 [46:39<10:53,  6.95s/it]

Modelo guardado - Recompensa media última 10: 203.50


 81%|████████▏ | 407/500 [46:45<10:21,  6.68s/it]

Modelo guardado - Recompensa media última 10: 201.00


 82%|████████▏ | 408/500 [46:51<09:42,  6.33s/it]

Modelo guardado - Recompensa media última 10: 186.00


 82%|████████▏ | 409/500 [46:58<09:53,  6.52s/it]

Modelo guardado - Recompensa media última 10: 176.50


 82%|████████▏ | 410/500 [47:07<10:51,  7.24s/it]

Modelo guardado - Recompensa media última 10: 182.50


 82%|████████▏ | 411/500 [47:13<10:06,  6.82s/it]

Modelo guardado - Recompensa media última 10: 151.00


 82%|████████▏ | 412/500 [47:21<10:44,  7.33s/it]

Modelo guardado - Recompensa media última 10: 173.50


 83%|████████▎ | 413/500 [47:29<10:46,  7.43s/it]

Modelo guardado - Recompensa media última 10: 193.00


 83%|████████▎ | 414/500 [47:41<12:42,  8.86s/it]

Modelo guardado - Recompensa media última 10: 207.00


 83%|████████▎ | 415/500 [47:47<11:16,  7.96s/it]

Modelo guardado - Recompensa media última 10: 201.00


 83%|████████▎ | 416/500 [47:56<11:40,  8.34s/it]

Modelo guardado - Recompensa media última 10: 222.50


 83%|████████▎ | 417/500 [48:05<11:35,  8.38s/it]

Modelo guardado - Recompensa media última 10: 251.50


 84%|████████▎ | 418/500 [48:11<10:34,  7.73s/it]

Modelo guardado - Recompensa media última 10: 258.00


 84%|████████▍ | 419/500 [48:16<09:29,  7.03s/it]

Modelo guardado - Recompensa media última 10: 255.00


 84%|████████▍ | 420/500 [48:30<12:08,  9.11s/it]

Modelo guardado - Recompensa media última 10: 253.50


 84%|████████▍ | 421/500 [48:33<09:41,  7.36s/it]

Modelo guardado - Recompensa media última 10: 245.00


 84%|████████▍ | 422/500 [48:43<10:23,  7.99s/it]

Modelo guardado - Recompensa media última 10: 232.50


 85%|████████▍ | 423/500 [48:46<08:31,  6.64s/it]

Modelo guardado - Recompensa media última 10: 209.50


 85%|████████▍ | 424/500 [48:59<10:34,  8.35s/it]

Modelo guardado - Recompensa media última 10: 215.50


 85%|████████▌ | 425/500 [49:07<10:23,  8.31s/it]

Modelo guardado - Recompensa media última 10: 235.50


 85%|████████▌ | 426/500 [49:15<10:02,  8.14s/it]

Modelo guardado - Recompensa media última 10: 244.50


 85%|████████▌ | 427/500 [49:18<08:10,  6.72s/it]

Modelo guardado - Recompensa media última 10: 208.50


 86%|████████▌ | 428/500 [49:28<09:06,  7.59s/it]

Modelo guardado - Recompensa media última 10: 205.50


 86%|████████▌ | 429/500 [49:36<09:14,  7.81s/it]

Modelo guardado - Recompensa media última 10: 221.00


 86%|████████▌ | 430/500 [49:45<09:21,  8.02s/it]

Modelo guardado - Recompensa media última 10: 210.00


 86%|████████▌ | 431/500 [49:54<09:38,  8.38s/it]

Modelo guardado - Recompensa media última 10: 226.00


 86%|████████▋ | 432/500 [50:02<09:35,  8.46s/it]

Modelo guardado - Recompensa media última 10: 220.50


 87%|████████▋ | 433/500 [50:14<10:24,  9.32s/it]

Modelo guardado - Recompensa media última 10: 275.50


 87%|████████▋ | 434/500 [50:21<09:41,  8.81s/it]

Modelo guardado - Recompensa media última 10: 257.00


 87%|████████▋ | 435/500 [50:29<09:01,  8.34s/it]

Modelo guardado - Recompensa media última 10: 255.00


 87%|████████▋ | 436/500 [50:36<08:29,  7.96s/it]

Modelo guardado - Recompensa media última 10: 234.50


 87%|████████▋ | 437/500 [50:41<07:35,  7.23s/it]

Modelo guardado - Recompensa media última 10: 247.50


 88%|████████▊ | 438/500 [50:52<08:28,  8.20s/it]

Modelo guardado - Recompensa media última 10: 247.00


 88%|████████▊ | 439/500 [51:02<09:08,  8.99s/it]

Modelo guardado - Recompensa media última 10: 259.00


 88%|████████▊ | 440/500 [51:11<08:46,  8.78s/it]

Modelo guardado - Recompensa media última 10: 255.50


 88%|████████▊ | 441/500 [51:19<08:35,  8.74s/it]

Modelo guardado - Recompensa media última 10: 267.50


 88%|████████▊ | 442/500 [51:25<07:26,  7.70s/it]

Modelo guardado - Recompensa media última 10: 258.00


 89%|████████▊ | 443/500 [51:32<07:16,  7.65s/it]

Modelo guardado - Recompensa media última 10: 221.50


 89%|████████▉ | 444/500 [51:36<06:09,  6.60s/it]

Modelo guardado - Recompensa media última 10: 212.00


 89%|████████▉ | 445/500 [51:43<06:09,  6.72s/it]

Modelo guardado - Recompensa media última 10: 202.50


 89%|████████▉ | 446/500 [51:48<05:25,  6.02s/it]

Modelo guardado - Recompensa media última 10: 188.50


 89%|████████▉ | 447/500 [51:54<05:25,  6.14s/it]

Modelo guardado - Recompensa media última 10: 192.00


 90%|████████▉ | 448/500 [52:01<05:35,  6.46s/it]

Modelo guardado - Recompensa media última 10: 195.00


 90%|████████▉ | 449/500 [52:10<06:01,  7.08s/it]

Modelo guardado - Recompensa media última 10: 174.50


 90%|█████████ | 450/500 [52:17<05:50,  7.01s/it]

Modelo guardado - Recompensa media última 10: 171.50


 90%|█████████ | 451/500 [52:27<06:25,  7.86s/it]

Modelo guardado - Recompensa media última 10: 178.50


 90%|█████████ | 452/500 [52:32<05:44,  7.17s/it]

Modelo guardado - Recompensa media última 10: 171.00


 91%|█████████ | 453/500 [52:40<05:51,  7.48s/it]

Modelo guardado - Recompensa media última 10: 183.00


 91%|█████████ | 454/500 [52:47<05:26,  7.10s/it]

Modelo guardado - Recompensa media última 10: 190.50


 91%|█████████ | 455/500 [52:54<05:27,  7.27s/it]

Modelo guardado - Recompensa media última 10: 201.50


 91%|█████████ | 456/500 [53:03<05:37,  7.68s/it]

Modelo guardado - Recompensa media última 10: 211.00


 91%|█████████▏| 457/500 [53:11<05:38,  7.87s/it]

Modelo guardado - Recompensa media última 10: 210.50


 92%|█████████▏| 458/500 [53:15<04:39,  6.65s/it]

Modelo guardado - Recompensa media última 10: 197.00


 92%|█████████▏| 459/500 [53:22<04:38,  6.79s/it]

Modelo guardado - Recompensa media última 10: 221.50


 92%|█████████▏| 460/500 [53:27<04:04,  6.12s/it]

Modelo guardado - Recompensa media última 10: 222.50


 92%|█████████▏| 461/500 [53:33<03:56,  6.07s/it]

Modelo guardado - Recompensa media última 10: 188.50


 92%|█████████▏| 462/500 [53:37<03:28,  5.49s/it]

Modelo guardado - Recompensa media última 10: 190.50


 93%|█████████▎| 463/500 [53:42<03:23,  5.51s/it]

Modelo guardado - Recompensa media última 10: 165.00


 93%|█████████▎| 464/500 [53:50<03:45,  6.27s/it]

Modelo guardado - Recompensa media última 10: 177.00


 93%|█████████▎| 465/500 [53:54<03:13,  5.53s/it]

Modelo guardado - Recompensa media última 10: 155.00


 93%|█████████▎| 466/500 [54:00<03:09,  5.58s/it]

Modelo guardado - Recompensa media última 10: 157.50


 93%|█████████▎| 467/500 [54:08<03:31,  6.41s/it]

Modelo guardado - Recompensa media última 10: 158.50


 94%|█████████▎| 468/500 [54:14<03:17,  6.16s/it]

Modelo guardado - Recompensa media última 10: 169.00


 94%|█████████▍| 469/500 [54:19<03:05,  5.98s/it]

Modelo guardado - Recompensa media última 10: 139.00


 94%|█████████▍| 470/500 [54:28<03:25,  6.83s/it]

Modelo guardado - Recompensa media última 10: 153.00


 94%|█████████▍| 471/500 [54:34<03:07,  6.48s/it]

Modelo guardado - Recompensa media última 10: 158.00


 94%|█████████▍| 472/500 [54:42<03:16,  7.02s/it]

Modelo guardado - Recompensa media última 10: 182.50


 95%|█████████▍| 473/500 [54:47<02:51,  6.33s/it]

Modelo guardado - Recompensa media última 10: 191.50


 95%|█████████▍| 474/500 [54:56<03:08,  7.25s/it]

Modelo guardado - Recompensa media última 10: 179.50


 95%|█████████▌| 475/500 [55:02<02:53,  6.94s/it]

Modelo guardado - Recompensa media última 10: 213.00


 95%|█████████▌| 476/500 [55:12<03:04,  7.69s/it]

Modelo guardado - Recompensa media última 10: 212.00


 95%|█████████▌| 477/500 [55:19<02:55,  7.62s/it]

Modelo guardado - Recompensa media última 10: 211.00


 96%|█████████▌| 478/500 [55:26<02:43,  7.44s/it]

Modelo guardado - Recompensa media última 10: 209.50


 96%|█████████▌| 479/500 [55:37<02:55,  8.38s/it]

Modelo guardado - Recompensa media última 10: 225.00


 96%|█████████▌| 480/500 [55:40<02:15,  6.80s/it]

Modelo guardado - Recompensa media última 10: 203.50


 96%|█████████▌| 481/500 [55:47<02:07,  6.71s/it]

Modelo guardado - Recompensa media última 10: 223.00


 96%|█████████▋| 482/500 [55:52<01:56,  6.45s/it]

Modelo guardado - Recompensa media última 10: 207.00


 97%|█████████▋| 483/500 [56:02<02:05,  7.36s/it]

Modelo guardado - Recompensa media última 10: 209.50


 97%|█████████▋| 484/500 [56:13<02:14,  8.43s/it]

Modelo guardado - Recompensa media última 10: 216.00


 97%|█████████▋| 485/500 [56:23<02:15,  9.04s/it]

Modelo guardado - Recompensa media última 10: 221.00


 97%|█████████▋| 486/500 [56:31<01:59,  8.56s/it]

Modelo guardado - Recompensa media última 10: 227.50


 97%|█████████▋| 487/500 [56:36<01:40,  7.70s/it]

Modelo guardado - Recompensa media última 10: 219.00


 98%|█████████▊| 488/500 [56:41<01:21,  6.81s/it]

Modelo guardado - Recompensa media última 10: 209.00


 98%|█████████▊| 489/500 [56:47<01:12,  6.62s/it]

Modelo guardado - Recompensa media última 10: 201.00


 98%|█████████▊| 490/500 [56:59<01:20,  8.04s/it]

Modelo guardado - Recompensa media última 10: 245.00


 98%|█████████▊| 491/500 [57:06<01:10,  7.78s/it]

Modelo guardado - Recompensa media última 10: 233.00


 98%|█████████▊| 492/500 [57:13<01:00,  7.59s/it]

Modelo guardado - Recompensa media última 10: 234.50


 99%|█████████▊| 493/500 [57:19<00:49,  7.02s/it]

Modelo guardado - Recompensa media última 10: 224.00


 99%|█████████▉| 494/500 [57:35<00:59,  9.85s/it]

Modelo guardado - Recompensa media última 10: 251.50


 99%|█████████▉| 495/500 [57:44<00:48,  9.70s/it]

Modelo guardado - Recompensa media última 10: 263.50


 99%|█████████▉| 496/500 [57:52<00:35,  8.93s/it]

Modelo guardado - Recompensa media última 10: 286.00


 99%|█████████▉| 497/500 [58:02<00:28,  9.42s/it]

Modelo guardado - Recompensa media última 10: 309.00


100%|█████████▉| 498/500 [58:08<00:16,  8.25s/it]

Modelo guardado - Recompensa media última 10: 320.50


100%|█████████▉| 499/500 [58:15<00:08,  8.10s/it]

Modelo guardado - Recompensa media última 10: 303.50


100%|██████████| 500/500 [58:24<00:00,  7.01s/it]

Modelo guardado - Recompensa media última 10: 273.00


In [ ]:
# Evaluación del agente en modo test
def test_agent(env, model, num_episodes=5):
    model.eval()
    total_rewards = []
    for _ in range(num_episodes):
        obs, _ = env.reset()
        state = stacker.reset(obs)
        done = False
        total_reward = 0
        while not done:
            action = select_action(model, state, epsilon=0.0, num_actions=num_actions, device=device)
            next_obs, reward, terminated, truncated, _ = env.step(action)
            state = stacker.step(next_obs)
            total_reward += reward
            done = terminated or truncated
        total_rewards.append(total_reward)
    avg_reward = np.mean(total_rewards)
    print(f'Recompensa media en test: {avg_reward:.2f}')
    return avg_reward

# Cargar modelo y evaluar
policy.load_state_dict(torch.load('modelo_final.pth'))
test_agent(env, policy)

Recompensa media en test: 285.00


285.0

### 3. Justificación de los parámetros seleccionados y de los resultados obtenidos


#### Arquitectura robusta para entornos visuales
El modelo Deep Q-Network (DQN) desarrollado utiliza una arquitectura de red neuronal convolucional que permite procesar de forma eficiente los estados visuales del entorno de juego Space Invaders. Las tres capas convolucionales extraen características espaciales de los frames, y las capas totalmente conectadas traducen dicha representación en valores Q por acción. 
Esta estructura ha demostrado ser efectiva para tareas de toma de decisiones en ambientes dinámicos y visualmente complejos.

#### Entrenamiento exitoso con mejora progresiva
Durante el entrenamiento, el agente logró mejorar su rendimiento de manera sostenida, alcanzando una recompensa media móvil superior a 270 puntos en varias iteraciones. Este crecimiento progresivo evidencia que el modelo aprendió una política efectiva mediante la retroalimentación obtenida del entorno, maximizando recompensas acumuladas a través de la exploración y explotación balanceadas.

#### Discusión de resultados
En este proyecto se implementó desde cero el algoritmo Deep Q-Network (DQN) para el entorno SpaceInvaders-v5 (equivalente moderno de SpaceInvaders-v0), utilizando una red neuronal convolucional como aproximador de la función Q.

Durante la fase de entrenamiento, se utilizaron técnicas clave como:
- Replay Buffer para romper la correlación entre muestras.
- Frame Stacking de 4 frames para incorporar memoria temporal.
- Target Network con actualizaciones periódicas para estabilizar el aprendizaje.
- Política ε-greedy con decaimiento exponencial para equilibrar exploración y explotación.

Tras 500 episodios de entrenamiento, se alcanzó una recompensa media de 285.00 puntos en modo test, calculada sobre 5 episodios sin exploración (ε = 0). Este resultado supera ampliamente el umbral mínimo de **20 puntos** requeridos para este proyecto.

Este buen desempeño se debe a:
- Una arquitectura CNN bien adaptada a entornos visuales.
- Uso efectivo de técnicas estabilizadoras del DQN original de DeepMind.
- Entrenamiento sostenido y uso de checkpoints para retomar progresos.

#### Conclusión
Este resultado indica que el agente no solo fue capaz de aprender una estrategia eficiente durante el entrenamiento, sino que también puede generalizar su comportamiento a nuevas partidas no vistas previamente, lo cual valida la robustez y aplicabilidad del enfoque DQN en escenarios reales de toma de decisiones.